### PR TITLE
[darwin] - implement proper ping method

### PR DESCRIPTION
**Pull request by Memphiz**
Thx to MrMC who did that one. This implementes a simple ping which can be used throughout all darwin platforms (osx, ios, tvos).

This will be needed for tvos because there are no system calls allowed. And in general it looks prettier having real code behind that ping function if you ask me.

This one only affects osx and ios...

### CODE REVIEW

**Comment by Memphiz** [comment]
jenkins build this please

**Comment by mk01** [comment]
@Memphiz @MrMC 
is there any special reason to go back to platform specific code / implementations and not just implement darwin/linux/bsd compatible variant ?

furthermore if it is not providing ipv6 support, thus requiring to re-add spawning ping6 as soon as the ipv6 cnetwork rewrite will be merged ?

**Comment by MrMC** [comment]
Are you going to code it ?

**Comment by Memphiz** [comment]
@mk01 - well the goal is to get rid of syscalls. Other then that i only do/want to do the backport of this. What will happen when ipv6 support is in and those ping methods only do ipv4? I think the former syscall also only does ipv4 and i think there is no ping6 tool on darwin either - so nothing lost at all maybe...

File: xbmc/network/osx/darwinICMPPing.m [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
++(void)pingHostAddress:(NSData*)hostAddress andResultCallback:(void(^)(NSString* result))result;
+@end
+
+int darwinICMPPingHost(const char *host, unsigned int timeout_ms)
+{
+  // also checks if host is a name or ip address
+  struct sockaddr_in callAddress = {0};
+  callAddress.sin_len = sizeof(callAddress);
+  callAddress.sin_family = AF_INET;
+  callAddress.sin_addr.s_addr = inet_addr(host);
```
Comment (t4-ravenbird) [inline xbmc/network/osx/darwinICMPPing.m]: Why not call darwinICMPPing(remoteIp, .. ) after lookup inet_addr and avoid duplicating a big chunk of code?

File: xbmc/network/osx/darwinICMPPing.m [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+  NSDate* _dateReference;
+}
+@property(nonatomic, strong) void(^resultCallback)(NSString* result);
+@end
+
+@implementation SimplePingClient
+
++(void)pingHostname:(NSString*)hostName andResultCallback:(void(^)(NSString* result))result
+{
+  static SimplePingClient* singletonPC = nil;
```
Comment (t4-ravenbird) [inline xbmc/network/osx/darwinICMPPing.m]: Is this singleton catching any replies for the duration of the ping? If ping is done by multiple threads is it asserted that the incoming replies is returned only to the correct caller?
(Disclaimer , I dont read applish)

File: xbmc/network/osx/darwinICMPPing.m (hunk 1 above)
Comment (MrMC) [inline xbmc/network/osx/darwinICMPPing.m]: one could do that.

**Comment by MrMC** [comment]
Don't get me wrong, ipv6 support would be nice but it's a major refactor to do it right. I also bet 90 percent use case puts kodi on a local ipv4 DHCP net.

In this case, the ONLY reason for this implementation of a native ping on iOS/tvOS is to provide a working WOA. 

Threading might be a issue, simple to fix, remove the singleton aspect and create/delete objects. Again this was aimed at providing a working WOA and that usage is serialized.

**Comment by t4-ravenbird** [comment]
I support the idea to avoid syscall and do it properly, so I support this PR in general. ipv6 is not supported other places also as far as I can tell so I can not see this is an issue.
But WOA will not serialize waking servers so mixing up ping-response can give nasty errors (2 different threads can try to wake 2 different servers at overlapping time .. each thread will block only until ping-response then carry on)

**Comment by MrMC** [comment]
beware there is lots of objc magic going on including being compiled with '-fobjc-arc'. I looked over the code again and I'm leaning to it not being a singleton. @Memphiz, your thoughts ?

**Comment by MrMC** [comment]
hahaha, this is one of those things where each time you look at it, you change your mind :) 
obj is a singleton but it's not quite clear about thread safely. It would be trivial to change and init/alloc the obj in darwinICMPPingHost/darwinICMPPing and ARC will take care of deleting it.

**Comment by Memphiz** [comment]
ehrm we don't use arc in mainline ...

**Comment by MrMC** [comment]
MrMC does, those files are tagged with '-fobjc-arc' in xcode.

MrMC.xcodeproj//project.pbxproj:		F5D5F2EC1C2083B700E3FE8B /* darwinICMPPing.m in Sources */ = {isa = PBXBuildFile; fileRef = F5D5F2EA1C2083B700E3FE8B /* darwinICMPPing.m */; settings = {COMPILER_FLAGS = "-fobjc-arc"; }; };

MrMC.xcodeproj//project.pbxproj:		F5D5F2F01C2083C600E3FE8B /* SimplePing.m in Sources */ = {isa = PBXBuildFile; fileRef = F5D5F2EE1C2083C600E3FE8B /* SimplePing.m */; settings = {COMPILER_FLAGS = "-fobjc-arc"; }; };

**Comment by Paxxi** [comment]
No opinion about the code but I'm curious, is this tested on tvos ? If syscalls are verboten my hunch would be that icmp could very well be as well, or binding to low ports.

**Comment by MrMC** [comment]
It has been in MrMC for iOS/tvOS for a month or so now, works like a champ. The whole point of this is to do a ping in a non-privileged way.

**Comment by MrMC** [comment]
I also suspect this method is limited to darwin. To quote Apple

"On most platforms ping requires privileges (it's implemented with a raw IP socket).  Mac OS X 10.2 and later, and all versions of iOS, include a special facility that allows you to ping without privileges.  Specifically, you can open a special, non-privileged ICMP socket that allows you to send and receive pings."

**Comment by Paxxi** [comment]
Thanks @davilla that answers it

**Comment by MartijnKaijser** [comment]
@Memphiz anything for v17 ?

**Comment by Memphiz** [comment]
No it needs more work...

**Comment by Rechi** [comment]
@Memphiz this needs a rebase

**Comment by Memphiz** [comment]
Closing this - mrmc switched back to system ping because this pr is broken...

**Comment by MrMC** [comment]
switch back to system ping for OSX, iOS/tvOS still uses it

**Pull request closed without merging**